typing/typedecl.ml

```diff
@@ -1370,53 +1370,68 @@ let transl_value_decl env loc valdecl =
     (fun () -> transl_value_decl env loc valdecl)
 
 (* Translate a "with" constraint -- much simplified version of
-    transl_type_decl. *)
-let transl_with_constraint env id row_path orig_decl sdecl =
-  Env.mark_type_used orig_decl.type_uid;
+   transl_type_decl. For a constraint [Sig with t = sdecl],
+   there are two declarations of interest in two environments:
+   - [sig_decl] is the declaration of [t] in [Sig],
+     in the environment [sig_env] (containing the declarations
+     of [Sig] before [t])
+   - [sdecl] is the new syntactic declaration, to be type-checked
+     in the current, outer environment [with_env].
+
+   In particular, note that [sig_env] is an extension of
+   [outer_env].
+*)
+let transl_with_constraint id row_path ~sig_env ~sig_decl ~outer_env sdecl =
+  Env.mark_type_used sig_decl.type_uid;
   reset_type_variables();
   Ctype.begin_def();
-  let tparams = make_params env sdecl.ptype_params in
+  let loc = sdecl.ptype_loc in
+  let tparams = make_params outer_env sdecl.ptype_params in
   let params = List.map (fun (cty, _) -> cty.ctyp_type) tparams in
-  let orig_decl = Ctype.instance_declaration orig_decl in
-  let arity_ok = List.length params = orig_decl.type_arity in
+  let sig_decl = Ctype.instance_declaration sig_decl in
+  let arity_ok = List.length params = sig_decl.type_arity in
   if arity_ok then
-    List.iter2 (Ctype.unify_var env) params orig_decl.type_params;
+    List.iter2 (Ctype.unify_var outer_env) params sig_decl.type_params;
```

Comment:
Actually, I think this should be `sig_env` rather than `outer_env`.
However, this means that this unification should only happen after all the `transl_simpl_type` (which should not refer to types of the signature), and protect against unification errors.